daml-script/dump/src/it/scala/com/daml/script/dump/IT.scala

```diff
@@ -118,63 +123,98 @@ final class IT
+      timeMode = ScriptTimeMode.WallClock,
+      initialClients = Participants(
+        default_participant = Some(new GrpcLedgerClient(client, ApplicationId("script"))),
+        participants = Map.empty,
+        party_participants = Map.empty,
+      ),
+    )
+  } yield ()
+
+  private def testOffset(
+      numParties: Int,
+      skip: Int,
   )(f: (LedgerClient, Seq[Ref.Party]) => Future[Unit]): Future[Assertion] =
     for {
       client <- LedgerClient(channel, clientConfiguration)
-      parties <- List
-        .range(0, numParties)
-        .traverse(_ => client.partyManagementClient.allocateParty(None, None).map(_.party))
+      parties <- allocateParties(client, numParties)
       // setup
       _ <- f(client, parties)
       before <- collectTrees(client, parties)
-      // build script dump
-      _ <- Main.run(
-        Config(
-          ledgerHost = "localhost",
-          ledgerPort = serverPort.value,
-          parties = parties,
-          outputPath = tmpDir,
-          damlScriptLib = damlScriptLib.toString,
-          sdkVersion = SdkVersion.sdkVersion,
-        )
-      )
-      // compile script dump
-      _ = Seq[String](
-        damlc.toString,
-        "build",
-        "--project-root",
-        tmpDir.toString,
-        "-o",
-        tmpDir.resolve("dump.dar").toString,
-      ).! shouldBe 0
-      // run script dump
-      newParties <- List
-        .range(0, numParties)
-        .traverse(_ => client.partyManagementClient.allocateParty(None, None).map(_.party))
-      encodedDar = DarReader().readArchiveFromFile(tmpDir.resolve("dump.dar").toFile).get
-      dar: Dar[(PackageId, Package)] = encodedDar
-        .map { case (pkgId, pkgArchive) => Decode.readArchivePayload(pkgId, pkgArchive) }
-      _ <- Runner.run(
-        dar,
-        Ref.Identifier(dar.main._1, Ref.QualifiedName.assertFromString("Dump:dump")),
-        inputValue = Some(JsArray(newParties.map(JsString(_)).toVector)),
-        timeMode = ScriptTimeMode.WallClock,
-        initialClients = Participants(
-          default_participant = Some(new GrpcLedgerClient(client, ApplicationId("script"))),
-          participants = Map.empty,
-          party_participants = Map.empty,
-        ),
-      )
+      // Reproduce ACS up to offset and transaction trees after offset.
+      offset =
+        if (skip == 0) { ledgerBegin }
+        else { ledgerOffset(before(skip - 1).offset) }
+      beforeCmp = before.drop(skip)
+      // reproduce from dump
+      dar <- createScriptDump(parties, offset)
+      newParties <- allocateParties(client, numParties)
+      _ <- runScriptDump(client, newParties, dar)
       // check that the new transaction trees are the same
       after <- collectTrees(client, newParties)
+      afterCmp = after.drop(after.length - beforeCmp.length)
     } yield {
-      TransactionEq.equivalent(before, after).fold(fail(_), _ => succeed)
+      TransactionEq.equivalent(beforeCmp, afterCmp).fold(fail(_), _ => succeed)
```

Comment:
Is it worth comparing the ACS before and after at `skip` as well?

Reply:
Yes, that would be a good check to have. However, unless I'm missing something, this would require a bit more infrastructure. In `before` the ACS up to the offset is created by the original create and exercise commands in their original order. In `after` it is created by plain creates in some topological ordering based on contract dependencies. There is no guarantee that the contracts are created in the same order. I'm not sure that current unification algorithm can handle this.